ksqldb-engine/src/main/java/io/confluent/ksql/physical/pull/HARouting.java

```diff
@@ -106,22 +107,39 @@ public void close() {
       final QueryId queryId,
       final PullQueryQueue pullQueryQueue
   ) {
-    final List<KsqlPartitionLocation> locations = pullPhysicalPlan.getMaterialization().locator()
+    final List<KsqlPartitionLocation> allLocations = pullPhysicalPlan.getMaterialization().locator()
         .locate(
             pullPhysicalPlan.getKeys(),
             routingOptions,
             routingFilterFactory
     );
 
-    final boolean anyPartitionsEmpty = locations.stream()
-        .anyMatch(location -> location.getNodes().isEmpty());
-    if (anyPartitionsEmpty) {
-      LOG.debug("Unable to execute pull query: {}. All nodes are dead or exceed max allowed lag.",
-                statement.getStatementText());
-      throw new MaterializationException("Unable to execute pull query. "
-          + "All nodes are dead or exceed max allowed lag.");
+    final Map<Integer, List<Host>> emptyPartitions = allLocations.stream()
+        .filter(loc -> loc.getNodes().stream().noneMatch(node -> node.getHost().isSelected()))
+        .collect(Collectors.toMap(
+            KsqlPartitionLocation::getPartition,
+            loc -> loc.getNodes().stream().map(KsqlNode::getHost).collect(Collectors.toList())));
+
+    if (!emptyPartitions.isEmpty()) {
+      final MaterializationException materializationException = new MaterializationException(
+          "Unable to execute pull query. "
+              + emptyPartitions.entrySet()
+              .stream()
+              .map(kv -> String.format(
+                  "Partition %s failed to find valid host. Hosts scanned: %s",
```

Comment:
nit: should we add an extra space after `Hosts scanned: %s ` to separate multiple partitions?

Reply:
I added the space the to `joining` collector so that the separator will be `", "` instead